src/materialized/src/http.rs

```diff
@@ -305,3 +351,43 @@ impl hyper::service::Service<Request<Body>> for ThirdPartyServer {
         }
     }
 }
+
+// Uses the HTTP Authorization header to fetch a JWT. Basic auth requires a
+// username and password that is "client,secret" that is exchanged for a JWT,
+// and whose username must match the JWT's email. Bearer auth can provide the
+// JWT directly. The JWT is validated and its email address is returned.
+async fn validate_http_mzcloud_authentication(
+    req: &Request<Body>,
+    mzcloud: &MzCloudAuthentication,
+) -> Result<String, anyhow::Error> {
+    let auth = match req.headers().get(AUTHORIZATION) {
+        Some(auth) => auth,
+        None => anyhow::bail!("no Authorization header"),
+    };
+    // There doesn't seem to be a way to get the Basic or Bearer struct out of
+    // the Authorization without doing a second decode. We can't just call those
+    // decodes first either, because they debug_assert that the HeaderValue starts
+    // with Basic or Bearer. Is there a better way to do this?
+    let (http_user, jwt): (Option<String>, String) =
+        if Authorization::<Basic>::decode(&mut once(auth)).is_ok() {
+            let basic = Basic::decode(auth).unwrap();
+            let jwt = mzcloud
+                .exchange_password_for_token(basic.password())
+                .await?;
+            (Some(basic.username().to_string()), jwt.access_token)
+        } else if Authorization::<Bearer>::decode(&mut once(auth)).is_ok() {
+            let bearer = Bearer::decode(auth).unwrap();
+            (None, bearer.token().to_string())
+        } else {
+            anyhow::bail!("unexpected authorization scheme");
+        };
```

Comment:
```suggestion
    let (http_user, jwt) = if let Some(basic) = req.headers().typed_get::<Authorization<Basic>>() {
        let jwt = mzcloud
            .exchange_password_for_token(basic.0.password())
            .await?;
        (Some(basic.0.username().to_string()), jwt.access_token)
    } else if let Some(basic) = req.headers().typed_get::<Authorization<Bearer>>() {
        (None, basic.0.token().to_string())
    } else {
        anyhow::bail!("invalid authorization");
    };
```

Reply:
Ahhh HeaderMapExt. Yes thanks. I also didn't realize that one can use `.entry()` which returns this interface instead of having to make a vec yourself. Rust can be obtuse if you don't happen to understand multiple traits at once.